doc/source/templates/04_finetuning_llms/finetune_llama.py

```diff
@@ -0,0 +1,438 @@
+import evaluate
+import torch
+import argparse
+import json
+import math
+import os
+import functools
+import time 
+import tree
+import pandas as pd
+from pathlib import Path
+import torch
+
+from accelerate import Accelerator, DeepSpeedPlugin
+from accelerate.utils import DummyOptim, DummyScheduler, set_seed
+from transformers import (
+    AutoModelForCausalLM,
+    get_linear_schedule_with_warmup,
+)
+
+
+import ray
+from ray import air
+from ray.air import session
+from ray.train.torch import TorchTrainer
+import ray.util.scheduling_strategies
+
+EVAL_BATCH_SIZE = 32
+MODEL = "tiiuae/falcon-7b"
+# MODEL = "tiiuae/falcon-40b"
+BLOCK_SIZE = 512
+OVERLAP_LENGTH = 128
+OPTIM_BETAS = (0.9, 0.999)
+OPTIM_EPS = 1e-8
+OPTIM_WEIGHT_DECAY = 0.
+
+def collate_fn(batch, tokenizer, device):
+    print('-'*80)
+    print(type(batch["input"]))
+    print('-'*80)
+    out_batch = tokenizer(
+        list(batch["input"]), 
+        padding="longest", 
+        max_length=BLOCK_SIZE, 
+        truncation=True, 
+        return_tensors="pt"    
+    )
+    out_batch["labels"] = out_batch["input_ids"].clone()
+    
+    out_batch = tree.map_structure(lambda x: x.to(device), out_batch)
+    
+    return out_batch
+
+
+def get_tokenizer(model_name, special_tokens):
+    
+    from transformers.models.llama.tokenization_llama import LlamaTokenizer
+
+    tokenizer = LlamaTokenizer.from_pretrained(model_name)
+    tokenizer.pad_token = tokenizer.eos_token
+    tokenizer.add_tokens(special_tokens, special_tokens=True)
+    
+    return tokenizer
+    
+    
+def get_ds(path):
+    # jsonl file
+    with open(path, 'r') as json_file:
+        items = [json.loads(x) for x in json_file]
+
+    dataset = {"input": []}
+    for item in items:
+        assert set(item.keys()) == {"input"}
+        dataset["input"].append(item["input"])
+    
+    df = pd.DataFrame.from_dict(dataset)
+    return ray.data.from_pandas(df)
+
+def evaluate(args, model, eval_dataloader, accelerator, eval_dataset):
+    model.eval()
+    losses = []
+    for step, batch in enumerate(eval_dataloader):
+        with torch.no_grad():
+            outputs = model(**batch)
+
+        loss = outputs.loss
+        losses.append(accelerator.gather(loss.repeat(EVAL_BATCH_SIZE)))
+
+    losses = torch.cat(losses)
+    losses = losses[: len(eval_dataset)]
+    try:
+        eval_loss = torch.mean(losses)
+        perplexity = math.exp(eval_loss)
+    except OverflowError:
+        perplexity = float("inf")
+    return perplexity, eval_loss
+
+def training_function(kwargs: dict):
+    os.environ["RANK"] = str(session.get_world_rank())
+    os.environ["WORLD_SIZE"] = str(session.get_world_size())
+    os.environ["LOCAL_RANK"] = str(session.get_local_rank())
```

Comment:
AIR TorchTrainer automatically sets these 3 environment variables. We can remove these lines here.